pkg/azuredisk/controllerserver.go

```diff
@@ -517,60 +518,152 @@ func (d *Driver) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
 			return nil, status.Errorf(codes.Aborted, "ListVolumes starting token(%d) can not be negative", start)
 		}
 	}
+	if d.cloud.KubeClient != nil && d.cloud.KubeClient.CoreV1() != nil && d.cloud.KubeClient.CoreV1().PersistentVolumes() != nil {
+		return d.listVolumesWithKubeClient(ctx, start, int(req.MaxEntries))
+	}
+	return d.listVolumesWithoutKubeClient(ctx, start, int(req.MaxEntries))
+}
 
-	disks, derr := d.cloud.DisksClient.ListByResourceGroup(ctx, d.cloud.ResourceGroup)
-	if derr != nil {
-		return nil, status.Error(codes.Internal, fmt.Sprintf("ListVolumes on rg(%s) failed with error: %v", d.cloud.ResourceGroup, derr.Error()))
+// listVolumesWithKubeClient is a helper function for ListVolumes used for when there is an available kubeclient
+func (d *Driver) listVolumesWithKubeClient(ctx context.Context, start, maxEntries int) (*csi.ListVolumesResponse, error) {
+	pvList, err := d.cloud.KubeClient.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "ListVolumes failed while fetching PersistentVolumes List with error: %v", err.Error())
+	}
+
+	// get all resource groups and put them into a sorted slice
+	rgMap := make(map[string]bool)
+	volSet := make(map[string]bool)
+	numVols := 0
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == "disk.csi.azure.com" {
+			diskURI := pv.Spec.CSI.VolumeHandle
+			if err := isValidDiskURI(diskURI); err != nil {
+				return nil, status.Errorf(codes.InvalidArgument, "disk URI(%s) is not valid: %v", diskURI, err)
+			}
+			rg, err := GetResourceGroupFromURI(diskURI)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "could not get resource group from diskURI(%s) with error(%v)", diskURI, err)
+			}
+			volSet[diskURI] = true
+			if _, visited := rgMap[rg]; visited {
+				continue
+			}
+			rgMap[rg] = true
+		}
+		numVols++
 	}
 
-	if start != 0 && start >= len(disks) {
+	if start != 0 && start > numVols {
 		return nil, status.Errorf(codes.Aborted, "ListVolumes starting token(%d) on rg(%s) is greater than total number of volumes", start, d.cloud.ResourceGroup)
 	}
+	resourceGroups := make([]string, len(rgMap))
+	i := 0
+	for rg := range rgMap {
+		resourceGroups[i] = rg
+		i++
+	}
+	sort.Strings(resourceGroups)
 
-	maxEntries := len(disks) - start
-	if req.MaxEntries > 0 && int(req.MaxEntries) < maxEntries {
-		maxEntries = int(req.MaxEntries)
+	// loop through each resourceGroup to get disk lists
+	entries := []*csi.ListVolumesResponse_Entry{}
+	numEntries, numVisited, numNotVisited := 0, 0, 0
+	for _, resourceGroup := range resourceGroups {
+		if maxEntries > 0 && numEntries >= maxEntries {
+			numNotVisited++
+			break
+		}
+		derr := d.listVolumesByResourceGroup(ctx, resourceGroup, &entries, start, maxEntries, &numVisited, &numNotVisited, &numEntries, &volSet)
```

Comment:
should only list volumes which is in `pvList.Items`

Reply:
I could still find etcd disks in `ListVolumes` logs, I think that's the root cause: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_azuredisk-csi-driver/725/pull-azuredisk-csi-driver-e2e-single-az/1359837206172143616/build-log.txt

Reply:
> should only list volumes which is in `pvList.Items`

I also thought that this was the cause and have added a `volSet` to keep track of the volumes in `pvList.Items` and only append the disk to the entry if it can be found in `volSet`. However, if etcd disks are still showing up in the results, I think the cause may be something else.